plugins/languages/lean.nix

```diff
@@ -0,0 +1,316 @@
+{
+  lib,
+  helpers,
+  config,
+  pkgs,
+  ...
+}:
+with lib; let
+  cfg = config.plugins.lean;
+in {
+  options.plugins.lean =
+    helpers.extraOptionsOptions
+    // {
+      enable = mkEnableOption "lean-nvim";
+
+      package = helpers.mkPackageOption "lean-nvim" pkgs.vimPlugins.lean-nvim;
+
+      leanPackage = mkOption {
+        type = with types; nullOr package;
+        default = pkgs.lean4;
+        description = "Which package to use for lean.";
+        example = null;
+      };
+
+      lsp =
+        helpers.defaultNullOpts.mkNullable
+        (
+          with types;
+            submodule {
+              freeformType = types.attrs;
+              options = {
+                enable = helpers.defaultNullOpts.mkBool true ''
+                  Whether to enable the Lean language server(s) ?
+
+                  Set to `false` to disable, otherwise should be a table of options to pass to
+                  `leanls`.
+                  See https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#leanls
+                  for details.
+                '';
+
+                on_attach = helpers.mkNullOrOption types.str ''
+                  The LSP handler.
+                '';
+
+                init_options = helpers.mkNullOrOption types.attrs ''
```

Comment:
Could you use types.attrsOf types.anything? 

They are better handled in merging:
```
    attrs = mkOptionType {
      merge = loc: foldl' (res: def: res // def.value) {};
    };
```

vs 

```
    attrsOf = elemType: mkOptionType rec {
      merge = loc: defs:
        mapAttrs (n: v: v.value) (filterAttrs (n: v: v ? value) (zipAttrsWith (name: defs:
            (mergeDefinitions (loc ++ [name]) elemType defs).optionalValue
          )
          # Push down position info.
          (map (def: mapAttrs (n: v: { inherit (def) file; value = v; }) def.value) defs)));
    };

    anything = mkOptionType {
      merge = loc: defs:
        let
          getType = value:
            if isAttrs value && isStringLike value
            then "stringCoercibleSet"
            else builtins.typeOf value;

          # Returns the common type of all definitions, throws an error if they
          # don't have the same type
          commonType = foldl' (type: def:
            if getType def.value == type
            then type
            else throw "The option `${showOption loc}' has conflicting option t
ypes in ${showFiles (getFiles defs)}"
          ) (getType (head defs).value) defs;

          mergeFunction = {
            # Recursively merge attribute sets
            set = (attrsOf anything).merge;
            # Safe and deterministic behavior for lists is to only accept one d
efinition
            # listOf only used to apply mkIf and co.
            list =
              if length defs > 1
              then throw "The option `${showOption loc}' has conflicting defini
tions, in ${showFiles (getFiles defs)}."
              else (listOf anything).merge;
            # This is the type of packages, only accept a single definition
            stringCoercibleSet = mergeOneOption;
            lambda = loc: defs: arg: anything.merge
              (loc ++ [ "<function body>" ])
              (map (def: {
                file = def.file;
                value = def.value arg;
              }) defs);
            # Otherwise fall back to only allowing all equal definitions
          }.${commonType} or mergeEqualOption;
        in mergeFunction loc defs;
  };
```

Reply:
Oh I didn't know that !
Done